{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%config InlineBackend.figure_format = 'retina'\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from tqdm import tqdm\n",
    "sns.set_palette('Dark2')\n",
    "sns.set_style({'axes.axisbelow': True, 'axes.edgecolor': '.15', 'axes.facecolor': 'white',\n",
    "               'axes.grid': True, 'axes.labelcolor': '.15', 'axes.linewidth': 1.25, \n",
    "               'figure.facecolor': 'white', 'font.family': ['sans-serif'], 'grid.color': '.15',\n",
    "               'grid.linestyle': ':', 'grid.alpha': .5, 'image.cmap': 'Greys', \n",
    "               'legend.frameon': False, 'legend.numpoints': 1, 'legend.scatterpoints': 1,\n",
    "               'lines.solid_capstyle': 'round', 'axes.spines.right': False, 'axes.spines.top': False,  \n",
    "               'text.color': '.15',  'xtick.top': False, 'ytick.right': False, 'xtick.color': '.15',\n",
    "               'xtick.direction': 'out', 'xtick.major.size': 6, 'xtick.minor.size': 3,\n",
    "               'ytick.color': '.15', 'ytick.direction': 'out', 'ytick.major.size': 6,'ytick.minor.size': 3})\n",
    "sns.set_context('talk')\n",
    "\n",
    "#http://phyletica.org/matplotlib-fonts/\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snapanalysis.config import OUTPUT_DIRECTORY as OUTPUT_DIRECTORY_MAIN\n",
    "\n",
    "OUTPUT_DIRECTORY = os.path.join(OUTPUT_DIRECTORY_MAIN, 'preprocessing')\n",
    "if not os.path.isdir(OUTPUT_DIRECTORY):\n",
    "    os.makedirs(OUTPUT_DIRECTORY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snapanalysis.preprocessing.cleanup.main import OUTPUT_FILE as ENRICHMENT_MODEL_INPUT\n",
    "from snapanalysis.models.enrichment.generate import RATIO_COLUMN\n",
    "from snapanalysis.models.enrichment.generate import rotate, EnrichmentDecoposition\n",
    "from snapanalysis.models.enrichment.generate import OUTPUT_FILE as ENRICHMENT_OUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snapanalysis.preprocessing.pulldown_metadata import OUTPUT_FILE as PD_META_OUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.HDFStore(PD_META_OUTPUT) as store:\n",
    "    pd_names = store['/meta/names_and_types']\n",
    "    pd_names = pd_names['Pull-Down name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Identification type\n",
      "Intensity\n",
      "Intensity H\n",
      "Intensity L\n",
      "Peptides\n",
      "Ratio H/L\n",
      "Ratio H/L (log2)\n",
      "Ratio H/L count\n",
      "Ratio H/L iso-count\n",
      "Ratio H/L normalized\n",
      "Ratio H/L normalized (log2)\n",
      "Ratio H/L type\n",
      "Ratio H/L variability [%]\n",
      "Razor + unique peptides\n",
      "Sequence coverage [%]\n",
      "Unique peptides\n"
     ]
    }
   ],
   "source": [
    "enrichment_data = pd.read_hdf(ENRICHMENT_OUTPUT, 'enrichment_data')\n",
    "pulldown_data = pd.read_hdf(ENRICHMENT_MODEL_INPUT, 'pulldown_data')\n",
    "\n",
    "protein_meta = pd.read_hdf(ENRICHMENT_MODEL_INPUT, 'protein_meta')\n",
    "maxquant_meta = pd.read_hdf(ENRICHMENT_MODEL_INPUT, 'maxquant_meta')\n",
    "\n",
    "flagged_proteins = pd.read_hdf(ENRICHMENT_MODEL_INPUT, 'flagged_proteins')\n",
    "\n",
    "print('\\n'.join(sorted(pulldown_data.columns)))\n",
    "\n",
    "pulldown_data = pulldown_data.unstack('Direction')\n",
    "\n",
    "pulldown_data.index = pulldown_data.index.swaplevel()\n",
    "enrichment_data.index = enrichment_data.index.swaplevel()\n",
    "\n",
    "pulldown_data = pulldown_data.sort_index()\n",
    "enrichment_data = enrichment_data.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enrichment\n",
      "Enrichment (imputed)\n",
      "Identification type (forward)\n",
      "Identification type (reverse)\n",
      "Imputation type\n",
      "Intensities missing\n",
      "Intensity (forward)\n",
      "Intensity (reverse)\n",
      "Intensity H (forward)\n",
      "Intensity H (reverse)\n",
      "Intensity L (forward)\n",
      "Intensity L (reverse)\n",
      "Missing data type\n",
      "Number of intensities seen\n",
      "Peptides (forward)\n",
      "Peptides (reverse)\n",
      "Ratio H/L (forward)\n",
      "Ratio H/L (log2) (forward)\n",
      "Ratio H/L (log2) (reverse)\n",
      "Ratio H/L (reverse)\n",
      "Ratio H/L count (forward)\n",
      "Ratio H/L count (reverse)\n",
      "Ratio H/L iso-count (forward)\n",
      "Ratio H/L iso-count (reverse)\n",
      "Ratio H/L normalized (forward)\n",
      "Ratio H/L normalized (log2) (adjusted, forward)\n",
      "Ratio H/L normalized (log2) (adjusted, imputed, forward)\n",
      "Ratio H/L normalized (log2) (adjusted, imputed, reverse)\n",
      "Ratio H/L normalized (log2) (adjusted, reverse)\n",
      "Ratio H/L normalized (log2) (forward)\n",
      "Ratio H/L normalized (log2) (imputed, forward)\n",
      "Ratio H/L normalized (log2) (imputed, reverse)\n",
      "Ratio H/L normalized (log2) (reverse)\n",
      "Ratio H/L normalized (reverse)\n",
      "Ratio H/L type (forward)\n",
      "Ratio H/L type (reverse)\n",
      "Ratio H/L variability [%] (forward)\n",
      "Ratio H/L variability [%] (reverse)\n",
      "Razor + unique peptides (forward)\n",
      "Razor + unique peptides (reverse)\n",
      "Residual\n",
      "Residual (imputed)\n",
      "Sequence coverage [%] (forward)\n",
      "Sequence coverage [%] (reverse)\n",
      "Unique peptides (forward)\n",
      "Unique peptides (reverse)\n",
      "p_value\n",
      "q_value\n",
      "significant\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(sorted(enrichment_data.columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio_columns_processed = [\n",
    "    'Ratio H/L normalized (log2) (adjusted, imputed, forward)',\n",
    "    'Ratio H/L normalized (log2) (adjusted, imputed, reverse)',\n",
    "    'Imputation type'\n",
    "]\n",
    "\n",
    "\n",
    "ratio_columns_unprocessed = [\n",
    "    'Ratio H/L normalized (log2) (forward)',\n",
    "    'Ratio H/L normalized (log2) (reverse)',\n",
    "    'Enrichment',\n",
    "    'Residual',\n",
    "]\n",
    "\n",
    "other_data_columns = [\n",
    "    'Identification type (forward)',\n",
    "    'Identification type (reverse)',\n",
    "    'Peptides (forward)',\n",
    "    'Peptides (reverse)',\n",
    "    'Unique peptides (forward)',\n",
    "    'Unique peptides (reverse)',\n",
    "    'Razor + unique peptides (forward)',\n",
    "    'Razor + unique peptides (reverse)',\n",
    "    'Intensity (forward)',\n",
    "    'Intensity (reverse)',\n",
    "    'Intensity H (forward)',\n",
    "    'Intensity H (reverse)',\n",
    "    'Intensity L (forward)',\n",
    "    'Intensity L (reverse)',\n",
    "    'Ratio H/L type (forward)',\n",
    "    'Ratio H/L type (reverse)',\n",
    "    'Ratio H/L count (forward)',\n",
    "    'Ratio H/L count (reverse)',\n",
    "    'Ratio H/L iso-count (forward)',\n",
    "    'Ratio H/L iso-count (reverse)',\n",
    "    'Ratio H/L variability [%] (forward)',\n",
    "    'Ratio H/L variability [%] (reverse)',\n",
    "    'Sequence coverage [%] (forward)',\n",
    "    'Sequence coverage [%] (reverse)'\n",
    "]\n",
    "\n",
    "protein_meta_columns = [\n",
    "    'Majority protein IDs', \n",
    "    'Gene names',\n",
    "    'Protein names'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "pull_downs = sorted(np.unique(enrichment_data.index.get_level_values(level='Pull-Down ID')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTPUT_FILENAME = os.path.join(OUTPUT_DIRECTORY, 'table-pulldowns.xlsx')\n",
    "OUTPUT_TSV_FILENAME_TEMPLATE = os.path.join(OUTPUT_DIRECTORY, 'table-pulldowns.sheet.{}.tsv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_red = '#7E3E4D'\n",
    "color_blue = '#41597E'\n",
    "color_white = '#F9F9F9'\n",
    "\n",
    "def conditional_format_dict(absmax=4, \n",
    "                            reverse=False,\n",
    "                            color_max=color_red,\n",
    "                            color_min=color_blue,\n",
    "                            color_mid=color_white): \n",
    "    \n",
    "    if reverse:\n",
    "        color_max, color_min = color_min, color_max\n",
    "        \n",
    "    \n",
    "    ans = {\n",
    "        'type': '3_color_scale',\n",
    "        'min_type': 'num',\n",
    "        'mid_type': 'num',\n",
    "        'max_type': 'num',\n",
    "        'min_color': color_min,\n",
    "        'mid_color': color_mid,\n",
    "        'max_color': color_max,\n",
    "        'min_value': -absmax,\n",
    "        'mid_value': 0,\n",
    "        'max_value': absmax,\n",
    "    }\n",
    "    \n",
    "    return ans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/site-packages/ipykernel_launcher.py:76: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H01\n",
      "H01M\n",
      "H02\n",
      "H03\n",
      "H03M\n",
      "H04\n",
      "H04M\n",
      "H05\n",
      "H06\n",
      "H07\n",
      "H07M\n",
      "H08\n",
      "H08M\n",
      "H09\n",
      "H10\n",
      "H11\n",
      "H12\n",
      "H13\n",
      "H14\n",
      "H15\n",
      "H16\n",
      "H17\n",
      "H18\n",
      "H19\n",
      "H20\n",
      "H21\n",
      "H22\n",
      "H23\n",
      "H24\n",
      "H25\n",
      "H26\n",
      "H27M\n",
      "H28\n",
      "H29\n",
      "H30\n",
      "H31\n",
      "H32\n",
      "H33\n",
      "H34\n",
      "H35\n",
      "H36\n",
      "H37\n",
      "H38\n",
      "H39\n",
      "H39M\n",
      "H40\n",
      "H41\n",
      "H42\n",
      "H43\n",
      "H44\n",
      "H45\n",
      "H46\n",
      "H46M\n",
      "H47\n",
      "H47M\n"
     ]
    }
   ],
   "source": [
    "\n",
    "writer = pd.ExcelWriter(OUTPUT_FILENAME, engine='xlsxwriter')\n",
    "\n",
    "workbook = writer.book\n",
    "\n",
    "border_type = 2\n",
    "\n",
    "format_header = workbook.add_format({\n",
    "    'bold': True,\n",
    "    'bottom': border_type,\n",
    "})\n",
    "\n",
    "format_bold = workbook.add_format({\n",
    "    'bold': True,\n",
    "})\n",
    "\n",
    "format_header_border_left = workbook.add_format({\n",
    "    'bold': True,\n",
    "    'bottom': border_type,\n",
    "    'left': border_type,\n",
    "})\n",
    "\n",
    "\n",
    "format_merged_header = workbook.add_format({\n",
    "    'bold': True,\n",
    "    'align': 'center',\n",
    "    'valign': 'vcenter',\n",
    "    'text_wrap': True,\n",
    "})\n",
    "\n",
    "format_merged_header_border_left = workbook.add_format({\n",
    "    'bold': True,\n",
    "    'align': 'center',\n",
    "    'valign': 'vcenter',\n",
    "    'text_wrap': True,\n",
    "    'left': border_type,\n",
    "})\n",
    "\n",
    "format_merged_header_border = workbook.add_format({\n",
    "    'bold': True,\n",
    "    'align': 'center',\n",
    "    'valign': 'vcenter',\n",
    "    'text_wrap': True,\n",
    "    'bottom': border_type,\n",
    "})\n",
    "\n",
    "\n",
    "format_merged_header_leftalign_bigger = workbook.add_format({\n",
    "    'bold': True,\n",
    "    'align': 'left',\n",
    "    'valign': 'vcenter',\n",
    "    'text_wrap': True,\n",
    "    'font_size': 12,\n",
    "})\n",
    "\n",
    "format_border_left = workbook.add_format({\n",
    "    'left': border_type,\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "index_sheet = workbook.add_worksheet('Index')\n",
    "\n",
    "index_sheet.write(0, 0, 'Sheet name', format_header)\n",
    "index_sheet.write(0, 1, 'Pull-Down', format_header)\n",
    "\n",
    "for row, (pull_down, pd_name) in enumerate(pd_names.iteritems(), start=1):\n",
    "    index_sheet.write(row, 0, pull_down)\n",
    "    index_sheet.write(row, 1, pd_name)\n",
    "    \n",
    "# Write also as TSV\n",
    "pd_names.to_csv(OUTPUT_TSV_FILENAME_TEMPLATE.format(\n",
    "    '{:02}.{}'.format(\n",
    "        1,\n",
    "        'index'\n",
    "    )),\n",
    "    sep='\\t',\n",
    ")\n",
    "    \n",
    "\n",
    "index_sheet.set_column(0, 0, 10, format_bold)\n",
    "index_sheet.set_column(1, 1, 60)\n",
    "\n",
    "for i, pull_down in enumerate(pull_downs, start=2):\n",
    "    \n",
    "    pd_name = pd_names.loc[pull_down]\n",
    "    sheet_name = pull_down\n",
    "    \n",
    "    \n",
    "    ed = enrichment_data.loc[pull_down]\n",
    "    order = ed.index\n",
    "\n",
    "    \n",
    "    output_pm = protein_meta.loc[order, protein_meta_columns].reset_index()\n",
    "    output_pm.index = order\n",
    "    \n",
    "    output_ratios = ed.loc[order, ratio_columns_processed]\n",
    "    output_ratios_unprocessed = ed.loc[order, ratio_columns_unprocessed]\n",
    "    output_ratios_other_data = ed.loc[order, other_data_columns]\n",
    "    \n",
    "    output_ratios['Imputation type'] = output_ratios['Imputation type'].fillna('none')\n",
    "    \n",
    "    print(pull_down)\n",
    "    \n",
    "    start_row = 3\n",
    "    start_col = 0\n",
    "    \n",
    "    last_row = start_row + len(order)\n",
    "\n",
    "    # Write data\n",
    "    \n",
    "    # -- Protein metadata --------\n",
    "    output_pm.to_excel(writer, \n",
    "                       startrow=start_row,\n",
    "                       startcol=start_col,\n",
    "                       header=False,\n",
    "                       index=False,\n",
    "                       sheet_name=sheet_name)\n",
    "    \n",
    "\n",
    "    # Need to do this after the first write with pandas otherise it doesn't work\n",
    "    worksheet = writer.sheets[sheet_name]\n",
    "    \n",
    "    # Protein headers\n",
    "    for col_number, label in enumerate(output_pm.columns, start=start_col):\n",
    "        worksheet.merge_range(start_row-2, col_number, start_row-1, col_number, label, \n",
    "                              format_merged_header_border) \n",
    "        \n",
    "    \n",
    "    worksheet.merge_range(start_row-3, start_col, \n",
    "                          start_row-3, start_col + len(output_pm.columns) - 1, \n",
    "                          f'{pull_down}: {pd_name}', \n",
    "                          format_merged_header_leftalign_bigger)\n",
    "    \n",
    "    # Change widths a bit\n",
    "    # Gene label\n",
    "    worksheet.set_column(start_col, start_col, 15, format_bold)\n",
    "    # Gene names\n",
    "    worksheet.set_column(start_col+2, start_col+2, 15)\n",
    "    # Protein names\n",
    "    worksheet.set_column(start_col+3, start_col+3, 30)\n",
    "    \n",
    "    # -- Main ratios -------------\n",
    "    first_ratio_col = start_col + len(output_pm.columns)\n",
    "    worksheet.set_column(first_ratio_col, first_ratio_col, None, format_border_left)\n",
    "    \n",
    "    output_ratios.to_excel(writer, \n",
    "                           startrow=start_row,\n",
    "                           startcol=first_ratio_col,\n",
    "                           header=False,\n",
    "                           index=False,\n",
    "                           sheet_name=sheet_name)\n",
    "    # Ratio headers\n",
    "    \n",
    "    # Have to hardcode headers here because the merge columns is tricky to implement\n",
    "    assert np.all(output_ratios.columns == ['Ratio H/L normalized (log2) (adjusted, imputed, forward)',\n",
    "                                            'Ratio H/L normalized (log2) (adjusted, imputed, reverse)',\n",
    "                                            'Imputation type'])\n",
    "    \n",
    "    \n",
    "    worksheet.merge_range(start_row-2, first_ratio_col, \n",
    "                          start_row-2, first_ratio_col + 2,\n",
    "                          'Ratio H/L normalized (log2)',\n",
    "                          format_merged_header_border_left)\n",
    "    names = {\n",
    "        'Ratio H/L normalized (log2) (adjusted, imputed, forward)': 'Forward',\n",
    "        'Ratio H/L normalized (log2) (adjusted, imputed, reverse)': 'Reverse',\n",
    "        'Imputation type': 'Imputation'\n",
    "    }\n",
    "    \n",
    "    for col_number, label in enumerate(output_ratios.columns, start=first_ratio_col):\n",
    "        label = names[label]\n",
    "        \n",
    "        if col_number == first_ratio_col:\n",
    "            format_ = format_header_border_left\n",
    "        else:\n",
    "            format_ = format_header\n",
    "            \n",
    "        worksheet.write(start_row-1, col_number, label, format_)    \n",
    "        \n",
    "    worksheet.merge_range(start_row-3, first_ratio_col, \n",
    "                          start_row-3, first_ratio_col + len(output_ratios.columns) - 1, \n",
    "                          'Post-processed data', format_merged_header_border_left)\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Ratio conditional formatting\n",
    "    \n",
    "    # Forward\n",
    "    worksheet.conditional_format(\n",
    "        start_row, first_ratio_col, \n",
    "        last_row, first_ratio_col,\n",
    "        conditional_format_dict()\n",
    "    )\n",
    "    \n",
    "    # Reverse\n",
    "    worksheet.conditional_format(\n",
    "        start_row, first_ratio_col + 1, \n",
    "        last_row, first_ratio_col + 1,\n",
    "        conditional_format_dict(reverse=True)\n",
    "    )\n",
    "    \n",
    "    # -- Raw ratios -------------------\n",
    "    \n",
    "    # Have to hardcode headers here because the merge columns is tricky to implement\n",
    "    assert np.all(output_ratios_unprocessed.columns == ['Ratio H/L normalized (log2) (forward)',\n",
    "                                                        'Ratio H/L normalized (log2) (reverse)',\n",
    "                                                        'Enrichment',\n",
    "                                                        'Residual'])\n",
    "    \n",
    "    first_ratio_unprocessed_col = first_ratio_col + len(output_ratios.columns)\n",
    "    worksheet.set_column(first_ratio_unprocessed_col, first_ratio_unprocessed_col, None, format_border_left)\n",
    "    \n",
    "    output_ratios_unprocessed.to_excel(writer, \n",
    "                                       startrow=start_row,\n",
    "                                       startcol=first_ratio_unprocessed_col,\n",
    "                                       header=False,\n",
    "                                       index=False,\n",
    "                                       sheet_name=sheet_name)\n",
    "    \n",
    "    # Raw ratio columns\n",
    "    \n",
    "    worksheet.merge_range(start_row-3, first_ratio_unprocessed_col, \n",
    "                          start_row-3, first_ratio_unprocessed_col + len(output_ratios_unprocessed.columns) - 1, \n",
    "                          'Raw data', format_merged_header_border_left)\n",
    "    \n",
    "    worksheet.merge_range(start_row-2, first_ratio_unprocessed_col, \n",
    "                         start_row-2, first_ratio_unprocessed_col + 1,\n",
    "                        'Ratio H/L normalized (log2)',\n",
    "                         format_merged_header_border_left)\n",
    "    \n",
    "    worksheet.merge_range(start_row-2, first_ratio_unprocessed_col + 2, \n",
    "                         start_row-2, first_ratio_unprocessed_col + 3,\n",
    "                        'Enrichment decomposition',\n",
    "                         format_merged_header)\n",
    "    \n",
    "    names = {\n",
    "        'Ratio H/L normalized (log2) (forward)': 'Forward',\n",
    "        'Ratio H/L normalized (log2) (reverse)': 'Reverse'\n",
    "    }\n",
    "    \n",
    "    for col_number, label in enumerate(output_ratios_unprocessed.columns, \n",
    "                                       start=first_ratio_unprocessed_col):\n",
    "        label = names.get(label, label)\n",
    "        \n",
    "        if col_number == first_ratio_unprocessed_col:\n",
    "            format_ = format_header_border_left\n",
    "        else:\n",
    "            format_ = format_header\n",
    "            \n",
    "        worksheet.write(start_row-1, col_number, label, format_)    \n",
    "        \n",
    "    \n",
    "    # -- Other useful data ---\n",
    "    first_other_data_col = first_ratio_unprocessed_col + len(output_ratios_unprocessed.columns)\n",
    "    worksheet.set_column(first_other_data_col, first_other_data_col, None, format_border_left)\n",
    "    \n",
    "    output_ratios_other_data.to_excel(writer, \n",
    "                                      startrow=start_row,\n",
    "                                      startcol=first_other_data_col,\n",
    "                                      header=False,\n",
    "                                      index=False,\n",
    "                                      sheet_name=sheet_name)\n",
    "    \n",
    "    ## Also save as TSV\n",
    "    output_ratios_processed_for_tsv = output_ratios.copy()\n",
    "    output_ratios_processed_for_tsv.columns = [f'Processed {c}' if c.startswith('Ratio') else c for c in output_ratios_processed_for_tsv.columns]\n",
    "    \n",
    "    output_ratios_unprocessed_for_tsv = output_ratios_unprocessed.copy()\n",
    "    output_ratios_unprocessed_for_tsv.columns = [f'Raw {c}' if c.startswith('Ratio') else c for c in output_ratios_unprocessed_for_tsv.columns]\n",
    "    \n",
    "    df_all_together_for_tsv = output_pm.join(output_ratios_processed_for_tsv).join(output_ratios_unprocessed_for_tsv).join(output_ratios_other_data)\n",
    "    \n",
    "    # Write also as TSV\n",
    "    df_all_together_for_tsv.to_csv(OUTPUT_TSV_FILENAME_TEMPLATE.format(\n",
    "        '{:02}.{}'.format(\n",
    "            i,\n",
    "            pull_down.lower()\n",
    "        )),\n",
    "        sep='\\t', index=False,\n",
    "    )\n",
    "\n",
    "    \n",
    "    # headers\n",
    "    worksheet.merge_range(start_row-3, first_other_data_col, \n",
    "                          start_row-3, first_other_data_col + len(output_ratios_other_data.columns) - 1, \n",
    "                          'Metadata', format_merged_header_border_left)\n",
    "    \n",
    "    cs = output_ratios_other_data.columns\n",
    "    \n",
    "    col_i = first_other_data_col\n",
    "    for forward_label, reverse_label in zip(cs[::2], cs[1::2]):\n",
    "        \n",
    "        assert forward_label.endswith('(forward)')\n",
    "        assert reverse_label.endswith('(reverse)')\n",
    "        \n",
    "        assert forward_label[:-len(' (forward)')] == reverse_label[:-len(' (reverse)')]\n",
    "        \n",
    "        common_label = forward_label[:-len(' (forward)')]\n",
    "        \n",
    "        \n",
    "        if col_i == first_other_data_col:\n",
    "            format_ = format_header_border_left\n",
    "            \n",
    "            fm_merged = format_merged_header_border_left\n",
    "        else:\n",
    "            format_ = format_header\n",
    "            fm_merged = format_merged_header\n",
    "        \n",
    "        worksheet.merge_range(start_row-2, col_i, \n",
    "                              start_row-2, col_i + 1,\n",
    "                              common_label,\n",
    "                              fm_merged)\n",
    "        \n",
    "        worksheet.write(start_row-1, col_i, 'Forward', format_) \n",
    "        worksheet.write(start_row-1, col_i+1, 'Reverse', format_header) \n",
    "        \n",
    "        col_i += 2\n",
    "        \n",
    "    worksheet.set_column(col_i, col_i, None, format_border_left)\n",
    "    \n",
    "    worksheet.autofilter(start_row-1, start_col, last_row, col_i-1)\n",
    "    worksheet.freeze_panes(start_row, first_ratio_unprocessed_col)\n",
    "#     break\n",
    "    \n",
    "    \n",
    "writer.save()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Ratio H/L normalized (log2) (adjusted, imputed, forward)',\n",
       "       'Ratio H/L normalized (log2) (adjusted, imputed, reverse)',\n",
       "       'Imputation type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_ratios.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Ratio H/L normalized (log2) (forward)',\n",
       "       'Ratio H/L normalized (log2) (reverse)', 'Enrichment', 'Residual'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_ratios_unprocessed.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene label</th>\n",
       "      <th>Majority protein IDs</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Protein names</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A087X222</th>\n",
       "      <td>A0A087X222</td>\n",
       "      <td>A0A087X222</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A0C4DFX4</th>\n",
       "      <td>A0A0C4DFX4</td>\n",
       "      <td>A0A0C4DFX4</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A0C4DGP2</th>\n",
       "      <td>A0A0C4DGP2</td>\n",
       "      <td>A0A0C4DGP2</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A0C4DGP5</th>\n",
       "      <td>A0A0C4DGP5</td>\n",
       "      <td>A0A0C4DGP5</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>F8VZ44;Q9NRG9;H3BU82</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>Aladin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATF</th>\n",
       "      <td>AATF</td>\n",
       "      <td>Q9NY61</td>\n",
       "      <td>AATF</td>\n",
       "      <td>Protein AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCF2</th>\n",
       "      <td>ABCF2</td>\n",
       "      <td>Q9UG63;C9JHK9;C9JZV3</td>\n",
       "      <td>ABCF2</td>\n",
       "      <td>ATP-binding cassette sub-family F member 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABHD12</th>\n",
       "      <td>ABHD12</td>\n",
       "      <td>Q8N2K0</td>\n",
       "      <td>ABHD12</td>\n",
       "      <td>Monoacylglycerol lipase ABHD12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABRAXAS1</th>\n",
       "      <td>ABRAXAS1</td>\n",
       "      <td>D6REL5;Q6UWZ7</td>\n",
       "      <td>ABRAXAS1</td>\n",
       "      <td>BRCA1-A complex subunit Abraxas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACD</th>\n",
       "      <td>ACD</td>\n",
       "      <td>A0A0C4DGT6;Q96AP0;R4GNJ5</td>\n",
       "      <td>ACD</td>\n",
       "      <td>Adrenocortical dysplasia protein homolog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACIN1 (1)</th>\n",
       "      <td>ACIN1 (1)</td>\n",
       "      <td>E7EQT4;Q9UKV3</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>Apoptotic chromatin condensation inducer in th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACIN1 (2)</th>\n",
       "      <td>ACIN1 (2)</td>\n",
       "      <td>S4R3H4</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACIN1 (3)</th>\n",
       "      <td>ACIN1 (3)</td>\n",
       "      <td>G3V3B0</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACIN1 (4)</th>\n",
       "      <td>ACIN1 (4)</td>\n",
       "      <td>G3V3T3</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSL3</th>\n",
       "      <td>ACSL3</td>\n",
       "      <td>O95573</td>\n",
       "      <td>ACSL3</td>\n",
       "      <td>Long-chain-fatty-acid--CoA ligase 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>P60709</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>Actin, cytoplasmic 1;Actin, cytoplasmic 1, N-t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTBL2</th>\n",
       "      <td>ACTBL2</td>\n",
       "      <td>Q562R1</td>\n",
       "      <td>ACTBL2</td>\n",
       "      <td>Beta-actin-like protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTG1</th>\n",
       "      <td>ACTG1</td>\n",
       "      <td>P63261</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>Actin, cytoplasmic 2;Actin, cytoplasmic 2, N-t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL6A</th>\n",
       "      <td>ACTL6A</td>\n",
       "      <td>O96019</td>\n",
       "      <td>ACTL6A</td>\n",
       "      <td>Actin-like protein 6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL8</th>\n",
       "      <td>ACTL8</td>\n",
       "      <td>Q9H568</td>\n",
       "      <td>ACTL8</td>\n",
       "      <td>Actin-like protein 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTR5</th>\n",
       "      <td>ACTR5</td>\n",
       "      <td>Q9H9F9</td>\n",
       "      <td>ACTR5</td>\n",
       "      <td>Actin-related protein 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTR6</th>\n",
       "      <td>ACTR6</td>\n",
       "      <td>Q9GZN1;F8W057;G3V1Y1;F8VSD1</td>\n",
       "      <td>ACTR6</td>\n",
       "      <td>Actin-related protein 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTR8</th>\n",
       "      <td>ACTR8</td>\n",
       "      <td>Q9H981</td>\n",
       "      <td>ACTR8</td>\n",
       "      <td>Actin-related protein 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACT[A1,A2,C1,G2]</th>\n",
       "      <td>ACT[A1,A2,C1,G2]</td>\n",
       "      <td>P68133;P68032;P63267;P62736;A6NL76</td>\n",
       "      <td>ACTA1;ACTC1;ACTG2;ACTA2</td>\n",
       "      <td>Actin, alpha skeletal muscle;Actin, alpha card...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAR</th>\n",
       "      <td>ADAR</td>\n",
       "      <td>P55265</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>Double-stranded RNA-specific adenosine deaminase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADNP</th>\n",
       "      <td>ADNP</td>\n",
       "      <td>Q9H2P0</td>\n",
       "      <td>ADNP</td>\n",
       "      <td>Activity-dependent neuroprotector homeobox pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AEBP2</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>G5EA50;Q6ZN18;F5GZR7</td>\n",
       "      <td>AEBP2</td>\n",
       "      <td>Zinc finger protein AEBP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFF1</th>\n",
       "      <td>AFF1</td>\n",
       "      <td>P51825</td>\n",
       "      <td>AFF1</td>\n",
       "      <td>AF4/FMR2 family member 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFF4</th>\n",
       "      <td>AFF4</td>\n",
       "      <td>Q9UHB7</td>\n",
       "      <td>AFF4</td>\n",
       "      <td>AF4/FMR2 family member 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFG3L2</th>\n",
       "      <td>AFG3L2</td>\n",
       "      <td>Q9Y4W6</td>\n",
       "      <td>AFG3L2</td>\n",
       "      <td>AFG3-like protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF544</th>\n",
       "      <td>ZNF544</td>\n",
       "      <td>M0QY11;J3KQC8;Q6NX49</td>\n",
       "      <td>ZNF544</td>\n",
       "      <td>Zinc finger protein 544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF580</th>\n",
       "      <td>ZNF580</td>\n",
       "      <td>Q9UK33</td>\n",
       "      <td>ZNF580</td>\n",
       "      <td>Zinc finger protein 580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF592</th>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Q92610;H0YM74</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Zinc finger protein 592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF593</th>\n",
       "      <td>ZNF593</td>\n",
       "      <td>O00488;A9J4F5</td>\n",
       "      <td>ZNF593</td>\n",
       "      <td>Zinc finger protein 593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF609</th>\n",
       "      <td>ZNF609</td>\n",
       "      <td>O15014</td>\n",
       "      <td>ZNF609</td>\n",
       "      <td>Zinc finger protein 609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF629</th>\n",
       "      <td>ZNF629</td>\n",
       "      <td>Q9UEG4</td>\n",
       "      <td>ZNF629</td>\n",
       "      <td>Zinc finger protein 629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF638</th>\n",
       "      <td>ZNF638</td>\n",
       "      <td>Q14966</td>\n",
       "      <td>ZNF638</td>\n",
       "      <td>Zinc finger protein 638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF644</th>\n",
       "      <td>ZNF644</td>\n",
       "      <td>Q9H582</td>\n",
       "      <td>ZNF644</td>\n",
       "      <td>Zinc finger protein 644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF652</th>\n",
       "      <td>ZNF652</td>\n",
       "      <td>Q9Y2D9;D6RF85</td>\n",
       "      <td>ZNF652</td>\n",
       "      <td>Zinc finger protein 652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF66</th>\n",
       "      <td>ZNF66</td>\n",
       "      <td>A0A0G2JMG9;Q6ZN08</td>\n",
       "      <td>ZNF66</td>\n",
       "      <td>Putative zinc finger protein 66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF668</th>\n",
       "      <td>ZNF668</td>\n",
       "      <td>Q96K58</td>\n",
       "      <td>ZNF668</td>\n",
       "      <td>Zinc finger protein 668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF687</th>\n",
       "      <td>ZNF687</td>\n",
       "      <td>Q8N1G0;H0Y5I5;F8WCX2</td>\n",
       "      <td>ZNF687</td>\n",
       "      <td>Zinc finger protein 687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF691</th>\n",
       "      <td>ZNF691</td>\n",
       "      <td>A0A0A0MRR1;Q5VV50;Q5VV52</td>\n",
       "      <td>ZNF691</td>\n",
       "      <td>Zinc finger protein 691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF710</th>\n",
       "      <td>ZNF710</td>\n",
       "      <td>Q8N1W2</td>\n",
       "      <td>ZNF710</td>\n",
       "      <td>Zinc finger protein 710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF740</th>\n",
       "      <td>ZNF740</td>\n",
       "      <td>Q8NDX6</td>\n",
       "      <td>ZNF740</td>\n",
       "      <td>Zinc finger protein 740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF768</th>\n",
       "      <td>ZNF768</td>\n",
       "      <td>H3BS42;Q9H5H4</td>\n",
       "      <td>ZNF768</td>\n",
       "      <td>Zinc finger protein 768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF787</th>\n",
       "      <td>ZNF787</td>\n",
       "      <td>A0A087WUD1;Q6DD87</td>\n",
       "      <td>ZNF787</td>\n",
       "      <td>Zinc finger protein 787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF808/ZNF845</th>\n",
       "      <td>ZNF808/ZNF845</td>\n",
       "      <td>Q96IR2;Q8N4W9</td>\n",
       "      <td>ZNF845;ZNF808</td>\n",
       "      <td>Zinc finger protein 845;Zinc finger protein 808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF839</th>\n",
       "      <td>ZNF839</td>\n",
       "      <td>A8K0R7;H0YM94</td>\n",
       "      <td>ZNF839</td>\n",
       "      <td>Zinc finger protein 839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNHIT1</th>\n",
       "      <td>ZNHIT1</td>\n",
       "      <td>O43257</td>\n",
       "      <td>ZNHIT1</td>\n",
       "      <td>Zinc finger HIT domain-containing protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRD1</th>\n",
       "      <td>ZNRD1</td>\n",
       "      <td>Q9P1U0</td>\n",
       "      <td>ZNRD1</td>\n",
       "      <td>DNA-directed RNA polymerase I subunit RPA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2</th>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>A0A0C4DGV5;O95218</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>Zinc finger Ran-binding domain-containing prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN21</th>\n",
       "      <td>ZSCAN21</td>\n",
       "      <td>Q9Y5A6;G3V1M0</td>\n",
       "      <td>ZSCAN21</td>\n",
       "      <td>Zinc finger and SCAN domain-containing protein 21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN26</th>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>A0A087WXM4;A0A087X2F1;A0A024RCN4;Q16670</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>Zinc finger and SCAN domain-containing protein 26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN29</th>\n",
       "      <td>ZSCAN29</td>\n",
       "      <td>Q8IWY8;A0A0C4DGM2</td>\n",
       "      <td>ZSCAN29</td>\n",
       "      <td>Zinc finger and SCAN domain-containing protein 29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN30</th>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>A0A0C4DGV0;Q86W11</td>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>Zinc finger and SCAN domain-containing protein 30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN31</th>\n",
       "      <td>ZSCAN31</td>\n",
       "      <td>Q96LW9;C9J423</td>\n",
       "      <td>ZSCAN31</td>\n",
       "      <td>Zinc finger and SCAN domain-containing protein 31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN5A</th>\n",
       "      <td>ZSCAN5A</td>\n",
       "      <td>A0A0C4DGQ1;Q9BUG6;K7EIX9;K7EMC4</td>\n",
       "      <td>ZSCAN5A</td>\n",
       "      <td>Zinc finger and SCAN domain-containing protein 5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWINT</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>O95229;A6NH27</td>\n",
       "      <td>ZWINT</td>\n",
       "      <td>ZW10 interactor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZ-type zinc finger-containing protein 3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1920 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene label                     Majority protein IDs  \\\n",
       "Gene label                                                                    \n",
       "A0A087X222              A0A087X222                               A0A087X222   \n",
       "A0A0C4DFX4              A0A0C4DFX4                               A0A0C4DFX4   \n",
       "A0A0C4DGP2              A0A0C4DGP2                               A0A0C4DGP2   \n",
       "A0A0C4DGP5              A0A0C4DGP5                               A0A0C4DGP5   \n",
       "AAAS                          AAAS                     F8VZ44;Q9NRG9;H3BU82   \n",
       "AATF                          AATF                                   Q9NY61   \n",
       "ABCF2                        ABCF2                     Q9UG63;C9JHK9;C9JZV3   \n",
       "ABHD12                      ABHD12                                   Q8N2K0   \n",
       "ABRAXAS1                  ABRAXAS1                            D6REL5;Q6UWZ7   \n",
       "ACD                            ACD                 A0A0C4DGT6;Q96AP0;R4GNJ5   \n",
       "ACIN1 (1)                ACIN1 (1)                            E7EQT4;Q9UKV3   \n",
       "ACIN1 (2)                ACIN1 (2)                                   S4R3H4   \n",
       "ACIN1 (3)                ACIN1 (3)                                   G3V3B0   \n",
       "ACIN1 (4)                ACIN1 (4)                                   G3V3T3   \n",
       "ACSL3                        ACSL3                                   O95573   \n",
       "ACTB                          ACTB                                   P60709   \n",
       "ACTBL2                      ACTBL2                                   Q562R1   \n",
       "ACTG1                        ACTG1                                   P63261   \n",
       "ACTL6A                      ACTL6A                                   O96019   \n",
       "ACTL8                        ACTL8                                   Q9H568   \n",
       "ACTR5                        ACTR5                                   Q9H9F9   \n",
       "ACTR6                        ACTR6              Q9GZN1;F8W057;G3V1Y1;F8VSD1   \n",
       "ACTR8                        ACTR8                                   Q9H981   \n",
       "ACT[A1,A2,C1,G2]  ACT[A1,A2,C1,G2]       P68133;P68032;P63267;P62736;A6NL76   \n",
       "ADAR                          ADAR                                   P55265   \n",
       "ADNP                          ADNP                                   Q9H2P0   \n",
       "AEBP2                        AEBP2                     G5EA50;Q6ZN18;F5GZR7   \n",
       "AFF1                          AFF1                                   P51825   \n",
       "AFF4                          AFF4                                   Q9UHB7   \n",
       "AFG3L2                      AFG3L2                                   Q9Y4W6   \n",
       "...                            ...                                      ...   \n",
       "ZNF544                      ZNF544                     M0QY11;J3KQC8;Q6NX49   \n",
       "ZNF580                      ZNF580                                   Q9UK33   \n",
       "ZNF592                      ZNF592                            Q92610;H0YM74   \n",
       "ZNF593                      ZNF593                            O00488;A9J4F5   \n",
       "ZNF609                      ZNF609                                   O15014   \n",
       "ZNF629                      ZNF629                                   Q9UEG4   \n",
       "ZNF638                      ZNF638                                   Q14966   \n",
       "ZNF644                      ZNF644                                   Q9H582   \n",
       "ZNF652                      ZNF652                            Q9Y2D9;D6RF85   \n",
       "ZNF66                        ZNF66                        A0A0G2JMG9;Q6ZN08   \n",
       "ZNF668                      ZNF668                                   Q96K58   \n",
       "ZNF687                      ZNF687                     Q8N1G0;H0Y5I5;F8WCX2   \n",
       "ZNF691                      ZNF691                 A0A0A0MRR1;Q5VV50;Q5VV52   \n",
       "ZNF710                      ZNF710                                   Q8N1W2   \n",
       "ZNF740                      ZNF740                                   Q8NDX6   \n",
       "ZNF768                      ZNF768                            H3BS42;Q9H5H4   \n",
       "ZNF787                      ZNF787                        A0A087WUD1;Q6DD87   \n",
       "ZNF808/ZNF845        ZNF808/ZNF845                            Q96IR2;Q8N4W9   \n",
       "ZNF839                      ZNF839                            A8K0R7;H0YM94   \n",
       "ZNHIT1                      ZNHIT1                                   O43257   \n",
       "ZNRD1                        ZNRD1                                   Q9P1U0   \n",
       "ZRANB2                      ZRANB2                        A0A0C4DGV5;O95218   \n",
       "ZSCAN21                    ZSCAN21                            Q9Y5A6;G3V1M0   \n",
       "ZSCAN26                    ZSCAN26  A0A087WXM4;A0A087X2F1;A0A024RCN4;Q16670   \n",
       "ZSCAN29                    ZSCAN29                        Q8IWY8;A0A0C4DGM2   \n",
       "ZSCAN30                    ZSCAN30                        A0A0C4DGV0;Q86W11   \n",
       "ZSCAN31                    ZSCAN31                            Q96LW9;C9J423   \n",
       "ZSCAN5A                    ZSCAN5A          A0A0C4DGQ1;Q9BUG6;K7EIX9;K7EMC4   \n",
       "ZWINT                        ZWINT                            O95229;A6NH27   \n",
       "ZZZ3                          ZZZ3                                   Q8IYH5   \n",
       "\n",
       "                               Gene names  \\\n",
       "Gene label                                  \n",
       "A0A087X222                                  \n",
       "A0A0C4DFX4                                  \n",
       "A0A0C4DGP2                                  \n",
       "A0A0C4DGP5                                  \n",
       "AAAS                                 AAAS   \n",
       "AATF                                 AATF   \n",
       "ABCF2                               ABCF2   \n",
       "ABHD12                             ABHD12   \n",
       "ABRAXAS1                         ABRAXAS1   \n",
       "ACD                                   ACD   \n",
       "ACIN1 (1)                           ACIN1   \n",
       "ACIN1 (2)                           ACIN1   \n",
       "ACIN1 (3)                           ACIN1   \n",
       "ACIN1 (4)                           ACIN1   \n",
       "ACSL3                               ACSL3   \n",
       "ACTB                                 ACTB   \n",
       "ACTBL2                             ACTBL2   \n",
       "ACTG1                               ACTG1   \n",
       "ACTL6A                             ACTL6A   \n",
       "ACTL8                               ACTL8   \n",
       "ACTR5                               ACTR5   \n",
       "ACTR6                               ACTR6   \n",
       "ACTR8                               ACTR8   \n",
       "ACT[A1,A2,C1,G2]  ACTA1;ACTC1;ACTG2;ACTA2   \n",
       "ADAR                                 ADAR   \n",
       "ADNP                                 ADNP   \n",
       "AEBP2                               AEBP2   \n",
       "AFF1                                 AFF1   \n",
       "AFF4                                 AFF4   \n",
       "AFG3L2                             AFG3L2   \n",
       "...                                   ...   \n",
       "ZNF544                             ZNF544   \n",
       "ZNF580                             ZNF580   \n",
       "ZNF592                             ZNF592   \n",
       "ZNF593                             ZNF593   \n",
       "ZNF609                             ZNF609   \n",
       "ZNF629                             ZNF629   \n",
       "ZNF638                             ZNF638   \n",
       "ZNF644                             ZNF644   \n",
       "ZNF652                             ZNF652   \n",
       "ZNF66                               ZNF66   \n",
       "ZNF668                             ZNF668   \n",
       "ZNF687                             ZNF687   \n",
       "ZNF691                             ZNF691   \n",
       "ZNF710                             ZNF710   \n",
       "ZNF740                             ZNF740   \n",
       "ZNF768                             ZNF768   \n",
       "ZNF787                             ZNF787   \n",
       "ZNF808/ZNF845               ZNF845;ZNF808   \n",
       "ZNF839                             ZNF839   \n",
       "ZNHIT1                             ZNHIT1   \n",
       "ZNRD1                               ZNRD1   \n",
       "ZRANB2                             ZRANB2   \n",
       "ZSCAN21                           ZSCAN21   \n",
       "ZSCAN26                           ZSCAN26   \n",
       "ZSCAN29                           ZSCAN29   \n",
       "ZSCAN30                           ZSCAN30   \n",
       "ZSCAN31                           ZSCAN31   \n",
       "ZSCAN5A                           ZSCAN5A   \n",
       "ZWINT                               ZWINT   \n",
       "ZZZ3                                 ZZZ3   \n",
       "\n",
       "                                                      Protein names  \n",
       "Gene label                                                           \n",
       "A0A087X222                                                      NaN  \n",
       "A0A0C4DFX4                                                      NaN  \n",
       "A0A0C4DGP2                                                      NaN  \n",
       "A0A0C4DGP5                                                      NaN  \n",
       "AAAS                                                         Aladin  \n",
       "AATF                                                   Protein AATF  \n",
       "ABCF2                    ATP-binding cassette sub-family F member 2  \n",
       "ABHD12                               Monoacylglycerol lipase ABHD12  \n",
       "ABRAXAS1                            BRCA1-A complex subunit Abraxas  \n",
       "ACD                        Adrenocortical dysplasia protein homolog  \n",
       "ACIN1 (1)         Apoptotic chromatin condensation inducer in th...  \n",
       "ACIN1 (2)                                                       NaN  \n",
       "ACIN1 (3)                                                       NaN  \n",
       "ACIN1 (4)                                                       NaN  \n",
       "ACSL3                           Long-chain-fatty-acid--CoA ligase 3  \n",
       "ACTB              Actin, cytoplasmic 1;Actin, cytoplasmic 1, N-t...  \n",
       "ACTBL2                                    Beta-actin-like protein 2  \n",
       "ACTG1             Actin, cytoplasmic 2;Actin, cytoplasmic 2, N-t...  \n",
       "ACTL6A                                        Actin-like protein 6A  \n",
       "ACTL8                                          Actin-like protein 8  \n",
       "ACTR5                                       Actin-related protein 5  \n",
       "ACTR6                                       Actin-related protein 6  \n",
       "ACTR8                                       Actin-related protein 8  \n",
       "ACT[A1,A2,C1,G2]  Actin, alpha skeletal muscle;Actin, alpha card...  \n",
       "ADAR               Double-stranded RNA-specific adenosine deaminase  \n",
       "ADNP              Activity-dependent neuroprotector homeobox pro...  \n",
       "AEBP2                                     Zinc finger protein AEBP2  \n",
       "AFF1                                       AF4/FMR2 family member 1  \n",
       "AFF4                                       AF4/FMR2 family member 4  \n",
       "AFG3L2                                          AFG3-like protein 2  \n",
       "...                                                             ...  \n",
       "ZNF544                                      Zinc finger protein 544  \n",
       "ZNF580                                      Zinc finger protein 580  \n",
       "ZNF592                                      Zinc finger protein 592  \n",
       "ZNF593                                      Zinc finger protein 593  \n",
       "ZNF609                                      Zinc finger protein 609  \n",
       "ZNF629                                      Zinc finger protein 629  \n",
       "ZNF638                                      Zinc finger protein 638  \n",
       "ZNF644                                      Zinc finger protein 644  \n",
       "ZNF652                                      Zinc finger protein 652  \n",
       "ZNF66                               Putative zinc finger protein 66  \n",
       "ZNF668                                      Zinc finger protein 668  \n",
       "ZNF687                                      Zinc finger protein 687  \n",
       "ZNF691                                      Zinc finger protein 691  \n",
       "ZNF710                                      Zinc finger protein 710  \n",
       "ZNF740                                      Zinc finger protein 740  \n",
       "ZNF768                                      Zinc finger protein 768  \n",
       "ZNF787                                      Zinc finger protein 787  \n",
       "ZNF808/ZNF845       Zinc finger protein 845;Zinc finger protein 808  \n",
       "ZNF839                                      Zinc finger protein 839  \n",
       "ZNHIT1                  Zinc finger HIT domain-containing protein 1  \n",
       "ZNRD1                   DNA-directed RNA polymerase I subunit RPA12  \n",
       "ZRANB2            Zinc finger Ran-binding domain-containing prot...  \n",
       "ZSCAN21           Zinc finger and SCAN domain-containing protein 21  \n",
       "ZSCAN26           Zinc finger and SCAN domain-containing protein 26  \n",
       "ZSCAN29           Zinc finger and SCAN domain-containing protein 29  \n",
       "ZSCAN30           Zinc finger and SCAN domain-containing protein 30  \n",
       "ZSCAN31           Zinc finger and SCAN domain-containing protein 31  \n",
       "ZSCAN5A           Zinc finger and SCAN domain-containing protein 5A  \n",
       "ZWINT                                               ZW10 interactor  \n",
       "ZZZ3                       ZZ-type zinc finger-containing protein 3  \n",
       "\n",
       "[1920 rows x 4 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_pm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
